File: implementations/python/sorting/cocktail/cocktail_sort.py
def cocktail_shaker_sort(values):
    a = list(values)
    start = 0
    end = len(a) - 1
    swapped = True
    while swapped:
        swapped = False
        for i in range(start, end):
            if a[i] > a[i + 1]:
                a[i], a[i + 1] = a[i + 1], a[i]
                swapped = True
        if not swapped:
            break
        swapped = False
        end -= 1
        for i in range(end, start, -1):
            if a[i - 1] > a[i]:
                a[i - 1], a[i] = a[i], a[i - 1]
                swapped = True
        start += 1
    return a

File: implementations/cpp/sorting/cocktail/cocktail_sort.cpp
#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <vector>

std::vector<int> cocktailShakerSort(std::vector<int> a) {
  int start = 0;
  int end = static_cast<int>(a.size()) - 1;
  bool swapped = true;
  while (swapped) {
    swapped = false;
    for (int i = start; i < end; ++i) if (a[i] > a[i + 1]) { std::swap(a[i], a[i + 1]); swapped = true; }
    if (!swapped) break;
    swapped = false;
    --end;
    for (int i = end; i > start; --i) if (a[i - 1] > a[i]) { std::swap(a[i - 1], a[i]); swapped = true; }
    ++start;
  }
  return a;
}
int main() {
  if (cocktailShakerSort({5, 1, 4, 2, 8, 0, -3}) != std::vector<int>({-3, 0, 1, 2, 4, 5, 8})) throw std::runtime_error("primary");
  if (cocktailShakerSort({3, 3, 2, 1}) != std::vector<int>({1, 2, 3, 3})) throw std::runtime_error("duplicate");
  std::cout << "cpp cocktail ok\n";
}
